apps/actions-dependencies-updater/src/workflows.mts

```diff
@@ -0,0 +1,401 @@
+   */
+  name?: string;
+  /**
+   * The workflow's filename, this is unique due to filesystem constraints.
+   */
+  file: string;
+  /**
+   * The path to the workflow file
+   */
+  path: string;
+  /**
+   * The workflow's jobs, and their dependencies as labelled encountered through a job's steps
+   */
+  jobs: Job[];
+}
+
+interface Job {
+  name: string;
+
+  containingWorkflow: string;
+
+  directDependencies: Action[];
+
+  indirectDependencies?: Action[];
+}
+
+/**
+ * Parses all workflows in the repository and returns them as a map of workflow name to workflow object.
+ * @param ctx the RunContext
+ * @returns a map of workflow name to workflow object, containing all workflows in the repository
+ */
+export async function parseWorkflows(ctx: RunContext) {
+  const workflowDirectory = join(ctx.repoDir, ".github", "workflows");
+  const workflowFiles = await listAllYamlFiles(workflowDirectory);
+
+  const workflows: WorkflowByName = {};
+
+  for (const workflowFilePath of workflowFiles) {
+    const workflow = await parseWorkflow(ctx, workflowFilePath);
+    workflows[workflow.file] = workflow;
+  }
+
+  return workflows;
+}
+
+/**
+ * Parses a workflow file and returns the workflow object
+ * @param ctx the RunContext
+ * @param workflow the name of the workflow
+ * @param path the path to the workflow file
+ * @returns the parsed workflow
+ */
+async function parseWorkflow(ctx: RunContext, path: string): Promise<Workflow> {
+  const file = basename(path);
+
+  log.debug(`Processing workflow: ${file}`);
+
+  const fileStr = await readFile(path, "utf-8");
+  const parsedFile = YAML.parse(fileStr) as WorkflowSchema;
+
+  if (parsedFile == null) {
+    log.warn(`Parsed as empty workflow file: ${file}`);
+    return { file, path, jobs: [] };
+  }
+
+  const { name, jobs: jobDefinitions } = parsedFile;
+
+  if (!jobDefinitions) {
+    log.warn(`No jobs found in workflow file: ${file}`);
+    return { name, file, path, jobs: [] };
+  }
+
+  const jobsPromises = Object.entries(jobDefinitions).map(
+    async ([jobKey, jobDefinition]) =>
+      parseJob(ctx, file, jobKey, jobDefinition),
+  );
+
+  const jobs = await Promise.all(jobsPromises);
+  return { name, file, path, jobs };
+}
+
+/**
+ * Parses a single job definition contained within a workflow.
+ * @param ctx the RunContext
+ * @param workflow the name of the workflow
+ * @param jobKey the key of the job
+ * @param jobDefinition the job definition
+ * @returns the parsed job
+ */
+async function parseJob(
+  ctx: RunContext,
+  workflow: string,
+  jobKey: string,
+  jobDefinition: WorkflowSchema["jobs"]["key"],
+): Promise<Job> {
+  log.debug(`Processing Job: ${workflow} - ${jobKey}`);
+
+  if ("uses" in jobDefinition) {
+    log.warn(
+      "Found Job which is uses a workflow_call, to initiate another workflow. Parsing this is not directly supported, but if the workflow is in the same repository it should be parsed regularly.",
```

Comment:
I don't really understand this log, could you clarify it?

Reply:
A workflow job can be another workflow definition (which has `workflow_call`) trigger. I don't really want to support that kind of relationship.

But if a job is essentially referencing another workflow, then that workflow it is referencing should be parsed and validated normally like all of the other workflows, and so we don't need to recurse from a workflow->job->workflow.

I don't know how to word that better? I'll throw it into a chatgpt and see I guess. This is also an edge-case I've seen only once.

Reply:
`"Job detected that triggers another workflow via 'workflow_call'. Direct parsing of this workflow-to-workflow relationship isn't supported. However, if the called workflow resides in the same repository, it will be parsed as a standard workflow."`